lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3616,10 +3620,12 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     return CheckedCastKind::ValueCast;
   }
 
-  // If the destination type can be a supertype of the source type, we are
-  // performing what looks like an upcast except it rebinds generic
-  // parameters.
-  if (fromType->isBindableTo(toType))
+  // If it is possible to substitute the generic arguments of source type
+  // with the destination generic archetypes, we are performing what looks
+  // like an upcast except it rebinds generic parameters. Or substitute
+  // destination to source generic archetypes, where we are performing
+  // what looks like an downcast except it also rebinds generic parameters.
```

Comment:
How about this?

```suggestion
  // We perform an upcast while rebinding generic parameters if it's possible
  // to substitute the generic arguments of the source type with the generic
  // archetypes of the destination type. Otherwise, if it's possible to substitute
  // the generic arguments of the destination type with the generic archetypes
  // of the source type, we perform a downcast instead.
```

Reply:
Looks better, thanks :)